src/material_library.cpp

```diff
@@ -0,0 +1,373 @@
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(char* filename) {
+  std::string fname(filename);
+  write_json(fname);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f;
+  f.open(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat) {
+  // if exists, get the material name from metadata make one instead
+  std::string mat_name;
+  int mat_number = -1;
+  std::set<int>::iterator mat_numb_it;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue ){
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+      mat.metadata["mat_number"] = mat_number;
+    }
+    mat_numb_it = mat_number_set.find(mat_number);
+    if (mat_numb_it != mat_number_set.end()) {
+      std::string msg = "The Material Number Conflict. Material Number ";
+      msg += mat_number;
+      msg += " is already in the library.";
+      warning(msg);
+    }
+  } 
+  pyne::matname_set::iterator key_it;
+  if (mat.metadata.isMember("name")) {
+    
+    if (Json::intValue <= mat.metadata["name"].type() &&
+        mat.metadata["name"].type() <= Json::realValue ){
+      mat_name = std::to_string(mat.metadata["name"].asInt());
+      mat.metadata["name"] = mat_name; 
+    } else {
+      mat_name = mat.metadata["name"].asString();
+    }
+    
+  } else {
+    if (mat_number == -1) {
+      mat_number = keylist.size(); //set a temp 
+    }
+    mat_name = "_" + std::to_string(mat_number);
+    mat.metadata["name"] = mat_name;
+  }
+  
+  add_material(mat_name, mat);  
+}
+
+void pyne::MaterialLibrary::add_material(char* key, pyne::Material mat) {
+  std::string key_name(key);
+  add_material(key_name, mat);
+}
+
+void pyne::MaterialLibrary::add_material(const std::string& key, pyne::Material mat) {
+  std::string mat_name;
+  int mat_number = -1;
+  std::set<int>::iterator mat_numb_it;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue ){
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+      mat.metadata["mat_number"] = mat_number;
+    }
+    mat_numb_it = mat_number_set.find(mat_number);
+    if (mat_numb_it != mat_number_set.end()) {
+    }
+    mat_numb_it = mat_number_set.find(mat_number);
+    if (mat_numb_it != mat_number_set.end()) {
+      std::string msg = "The Material Number Conflict. Material Number ";
+      msg += mat_number;
+      msg += " is already in the library.";
+      warning(msg);
+    }
```

Comment:
This is a repeated block of code. Can we extract it into a separate method maybe?

Reply:
I simplified it and removed the test on the first one as it will call the second one at the end of the method.